controllers/appwrapper_controller.go

```diff
@@ -92,62 +79,100 @@ const (
 func (r *AppWrapperReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 
 	_ = log.FromContext(ctx)
-
 	var appwrapper arbv1.AppWrapper
+
 	if err := r.Get(ctx, req.NamespacedName, &appwrapper); err != nil {
 		if apierrors.IsNotFound(err) {
 			//ignore not-found errors, since we can get them on delete requests.
 			return ctrl.Result{}, nil
 		}
 		klog.Error(err, "unable to fetch appwrapper")
 	}
-
-	factory := machineinformersv1beta1.NewSharedInformerFactoryWithOptions(machineClient, resyncPeriod()(), machineinformersv1beta1.WithNamespace(""))
-	informer := factory.Machine().V1beta1().MachineSets().Informer()
-	msLister = factory.Machine().V1beta1().MachineSets().Lister()
-	machineLister = factory.Machine().V1beta1().Machines().Lister()
-
 	// todo: Move the getOCMClusterID call out of reconcile loop.
 	// Only reason we are calling it here is that the client is not able to make
 	// calls until it is started, so SetupWithManager is not working.
 	if !useMachineSets && ocmClusterID == "" {
 		getOCMClusterID(r)
 	}
 
-	stopper := make(chan struct{})
-	defer close(stopper)
-	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    onAdd,
-		UpdateFunc: onUpdate,
-		DeleteFunc: onDelete,
-	})
-	go informer.Run(stopper)
-	if !cache.WaitForCacheSync(stopper, informer.HasSynced) {
-		klog.Info("Wait for cache to sync")
+	// Checks for finalizer on appwrapper where its deletion timestamp == 0
+	if appwrapper.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&appwrapper, finalizerName) {
+			//onAdd replacement
+			if appwrapper.Status.State == arbv1.AppWrapperStateEnqueued || appwrapper.Status.State == "" {
```

Comment:
Here, the logic of scaling up is guarded by the finalizer, but there is still a possibility that the update to add the finalizer fails, which leads to duplicated scaling. There is the `scaledAppwrapper` global variable, that holds an in-memory state of the scaled AppWrappers, but it's also fragile, as it's not persisted, for example in case the controller restarted.

I'd suggest to decouple the finalizer logic from the scaling invariant guard, and implement the later by relying on the persisted state in the cluster, e.g., by checking the existence of MachinePool / MachineSet with the corresponding AppWrapper label(s), that were atomically added during the creation of these resources.